### ISSUE TITLE
CLI: www/bin/magento is not using the correct error handler

### ISSUE THREAD

**Issue by heldchen** [magento/magento2#5190]
## Steps to reproduce
1. Install Magento from `develop` branch.
2. Use `www/bin/magento`
## Expected result
1. `Magento\Framework\App\ErrorHandler` is used as error handler as intented in `www/bin/magento`
## Actual result
1. `Composer\Util\ErrorHandler` is used as error handler instead

the problem is that in `www/bin/magento` set_error_handler is called before instantiating the `Magento\Framework\Console\Cli` class:

```
try {
    $handler = new \Magento\Framework\App\ErrorHandler();
    set_error_handler([$handler, 'handler']);

    $application = new Magento\Framework\Console\Cli('Magento CLI', AppInterface::VERSION);
```

during the instantiation, the previously error handler is overwritten and `Composer\Util\ErrorHandler` used: `Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create()` calls `Composer\Console\Application->__construct()` which calls `Composer\Util\ErrorHandler::register()`. unfortunately, this handler is treating `E_NOTICE` warnings as exceptions, whereas the Magento error handler has a more sane default.

to mitigate the issue, a (ugly) workaround can be to be re-set the error handler to the intended handler after instantiating the Cli class:

```
try {
    $handler = new \Magento\Framework\App\ErrorHandler();
    set_error_handler([$handler, 'handler']);

    $application = new Magento\Framework\Console\Cli('Magento CLI', AppInterface::VERSION);
    set_error_handler([$handler, 'handler']);
```

**Comment by heldchen** [comment]
problem still present in 2.1.1

**Comment by bh-ref** [comment]
problem still present in 2.1.4

**Comment by magento-engcom-team** [comment]
@heldchen, thank you for your report.
The issue is already fixed in 2.2.0
